mars/dataframe/reduction/core.py

```diff
@@ -849,6 +851,23 @@ def add_function(self, func, ndim, cols=None, func_name=None):
             func_name = f"<custom_{self._custom_counter}>"
             self._custom_counter += 1
 
+        raw_func_name = getattr(func, "__name__", None)
+        if (
+            raw_func_name is not None
+            and raw_func_name != "nunique"
+            and not isinstance(func, functools.partial)
+        ):
+            if not (
+                hasattr(func, "__globals__")
+                and "__name__" in func.__globals__
+                and func.__globals__["__name__"].startswith("mars")
+            ):
```

Comment:
These conditions look too complex, handle builtin functions and Numpy functions separately.
``` Python
if inspect.isbuiltin(func):
    ...
elif func.__module__.split('.')[0] == 'numpy':
    ...
```